File: LeetCode/remote/unique_binary_search_trees.cs
// Given n, how many structurally unique BST's (binary search trees) that store values 1...n?
//
// For example,
// Given n = 3, there are a total of 5 unique BST's.
//
//    1         3     3      2      1
//     \       /     /      / \      \
//      3     2     1      1   3      2
//     /     /       \                 \
//    2     1         2                 3
//

using System;

public class Solution {
    public int NumTrees(int n) {
        if (n < 2) 
        {
            return 1;
        }
        
        var count = 0;
        for (var i = 1; i <= n; i++)
        {
            count += NumTrees(i - 1) * NumTrees(n - i);
        }

        return count;
    }

    static void Main()
    {
        var c = new Solution();
        Console.WriteLine(c.NumTrees(3));
        Console.WriteLine(c.NumTrees(19));
    }
}

File: LeetCode/remote/unique_binary_search_trees.rb
#   https://leetcode.com/problems/unique-binary-search-trees/
#
#   Given n, how many structurally unique BST's (binary search trees) that store values 1...n?
#
#   For example,
#   Given n = 3, there are a total of 5 unique BST's. 

#   TLE
def num_trees(n)
    trees(1, n)
end

def trees(low, high)
    low >= high ? 1 : low.upto(high).map { |x| trees(low, x - 1) * trees(x + 1, high) }.inject(:+)
end
